tests/integration/pilot/mirror_test.go

```diff
@@ -0,0 +1,248 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package pilot
+
+import (
+	"fmt"
+	"math"
+	"strings"
+	"sync"
+	"testing"
+	"time"
+
+	"istio.io/istio/pkg/test/util/retry"
+	"istio.io/istio/tests/util"
+
+	envoyAdmin "github.com/envoyproxy/go-control-plane/envoy/admin/v2alpha"
+	"github.com/hashicorp/go-multierror"
+
+	"istio.io/istio/pkg/test/framework"
+	"istio.io/istio/pkg/test/framework/components/echo"
+	"istio.io/istio/pkg/test/framework/components/echo/echoboot"
+	"istio.io/istio/pkg/test/framework/components/environment"
+	"istio.io/istio/pkg/test/framework/components/namespace"
+	"istio.io/istio/pkg/test/util/file"
+	"istio.io/istio/pkg/test/util/structpath"
+	"istio.io/istio/pkg/test/util/tmpl"
+)
+
+//	Virtual service topology
+//
+//	    a                      b                     c
+//	|-------|             |-------|    mirror   |-------|
+//	| Host0 | ----------> | Host1 | ----------> | Host2 |
+//	|-------|             |-------|             |-------|
+//
+
+type VirtualServiceMirrorConfig struct {
+	Name      string
+	Namespace string
+	Absent    bool
+	Percent   float64
+}
+
+type testCaseMirror struct {
+	name       string
+	absent     bool
+	percentage float64
+	threshold  float64
+}
+
+func TestMirroring(t *testing.T) {
+	cases := []testCaseMirror{
+		{
+			name:       "mirror-percent-absent",
+			absent:     true,
+			percentage: 100.0,
+			threshold:  0.0,
+		},
+		{
+			name:       "mirror-50",
+			percentage: 50.0,
+			threshold:  10.0,
+		},
+		{
+			name:       "mirror-10",
+			percentage: 10.0,
+			threshold:  5.0,
+		},
+		{
+			name:       "mirror-80",
+			percentage: 80.0,
+			threshold:  10.0,
+		},
+		{
+			name:       "mirror-0",
+			percentage: 0.0,
+			threshold:  0.0,
+		},
+	}
+
+	framework.
+		NewTest(t).
+		RequiresEnvironment(environment.Kube).
```

Comment:
Just wondering why? I saw you mention the traffic shifting one doesn't either, maybe need some fixes for the framework

Reply:
Yeah, it does not work on native as well. We should probably create a bug and investigate that. It could be done separately from this PR IMO.